[FrameworkBundle] Generate `Config` class

alexandre-daubois opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -
| License       | MIT

This PR adds a new `Config` class to configure a Symfony application:

```php
<?php

/** @var \Symfony\Config\Config $config */
$config->framework([
    'secret' => 'abc123',
]);

$config->parameters([
    'some_parameter' => 'some_value',
]);

$config
    ->imports(['test.php'])
    ->services([
        ['id' => 'some.service', 'class' => \App\Service\ConcreteService::class],
    ]);
```

One method exists for each extension, as well as `services()`, `imports()` and `parameters()`.

The class is generated at cache warmup, allowing to generate the corresponding phpdoc. Here a sneak peek of what the functions looks like:

```php
<?php

namespace Symfony\Config;

/**
 * This trait is automatically generated to help in creating a config.
 */
trait WebProfilerTrait
{
    /**
     * @param array{
     *     toolbar?: array{ // Profiler toolbar configuration
     *         enabled?: bool, // Default: false
     *         ajax_replace?: bool, // Replace toolbar on AJAX requests Default: false
     *     },
     *     intercept_redirects?: bool, // Default: false
     *     excluded_ajax_paths?: string|int|float|bool, // Default: "^/((index|app(_[\\w]+)?)\\.php/)?_wdt"
     * } $config
     */
    public function webProfiler(array $config): static
    {
        $this->webProfilerConfig->configure($config);

        return $this;
    }
}
```

This will bring nice IDE autocomplete as well as strong static analysis. Also, you're always one click away from the complete documentation.

---

Credits to Nicolas, Kevin and Ryan for the idea 🙂

alexandre-daubois committed:
[FrameworkBundle] Generate configuration classes and traits

stof commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,180 @@
+        }
+
+        if (self::FORMAT_JETBRAINS_ATTRIBUTE === $format) {
+            return static::doGenerateJetBrainsArrayShape($node);
```
using `static::` to reference private static methods does not make sense (and would be broken if the class was not final). Use `self::` instead.

stof commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,180 @@
+        $arrayShape = 'array';
+
+        if (!$children = $node->getChildren()) {
+            return $arrayShape.'<array-key, mixed>';
```
This does not seem to handle prototyped array nodes properly, as those often don't accept `mixed`. And a big part of the configuration settings of Symfony are nested inside a prototyped node (the full SecurityBundle configuration for instance)

stof replied:
btw, this defeats the argument about providing strong static analysis, which means it makes this strictly inferior to the existing generated fluent API for now.

alexandre-daubois commented:
Thank you for your review! I'm having a look at it and will soon address your comments

alexandre-daubois commented:
Prototyped arrays are now supported, here's a screenshot from the newly generated config (snippet):

![image](https://github.com/user-attachments/assets/843564c8-e379-47f9-bca2-1fd8d009ee0b)

Providers, password hashers, etc. are all prototypes.

dunglas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,214 @@
+            return 'array<'.self::handleNodeType($prototype).'>';
+        }
+
+        $arrayShape = 'array{'.\PHP_EOL;
```
```suggestion
        $arrayShape = "array{\n";
```

It's better to always use the UNIX line terminator for PHP scripts. It also works on Windows and allows generating the exact same files across OSes. This will also prevent mixing UNIX and Windows end of line because some UNIX ones are hardcoded later.

alexandre-daubois replied:
I wasn't aware of this. Thanks, that's updated!

dunglas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,214 @@
+                $arrayShape .= $handleNode($child, $nestingLevel);
+            }
+
+            $arrayShape .= ','.\PHP_EOL;
```
```suggestion
            $arrayShape .= ",\n";
```

dunglas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,214 @@
+            return 'array<'.self::handleNodeType($prototype).'>';
+        }
+
+        $shape .= '['.\PHP_EOL;
```
```suggestion
        $shape .= "[\n";
```

dunglas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,214 @@
+                return self::doGenerateJetBrainsArrayShape($node, 1 + $nestingLevel);
+            } else {
+                return "'".self::handleNodeType($node)."'";
+            }
```
```suggestion
            }

            return "'".self::handleNodeType($node)."'";
```

dunglas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,214 @@
+                $shape .= $handleNode($child, $nestingLevel);
+            }
+
+            $shape .= ','.self::generateInlinePhpDocForNode($child).\PHP_EOL;
```
```suggestion
            $shape .= ','.self::generateInlinePhpDocForNode($child)."\n";
```

dunglas commented on src/Symfony/Component/Config/Builder/ConfigBuilderGenerator.php:
```diff
@@ -102,6 +120,29 @@ private function writeClasses(): void
 
+    private function buildConfigureOption(ArrayNode $node, ClassBuilder $class): void
+    {
+        $class->addUse('JetBrains\PhpStorm\ArrayShape');
```
```suggestion
        $class->addUse(JetBrains\PhpStorm\ArrayShape::class);
```

(this works even if the dependency isn't installed)

dunglas commented on src/Symfony/Component/Config/Builder/ConfigBuilderGenerator.php:
```diff
@@ -469,10 +510,24 @@ private function getSingularName(PrototypedArrayNode $node): string
-        $body = '$output = [];';
+        $body = '';
+        if ($rootClass) {
+            $body = 'if ($this->configOutput) {
```
Maybe use nowdoc syntax for readability? This also unlocks highlighting in some IDEs (PhpStorm) if the identifier is the language name.

alexandre-daubois replied:
Unfortunately not possible here, because we need some spacing for `$output` to be added in the case of `$rootClass` being true. With nowdoc syntax, the four spaces are automatically removed because there's not character after. But I did the update at your other suggestion's place 🙂

dunglas commented on src/Symfony/Component/Config/Builder/ConfigFunctionGenerator.php:
```diff
@@ -0,0 +1,84 @@
+            foreach ($configurations as $alias => $configuration) {
+                $function->addParam('array', $alias);
+                $function->addParamDefaultValue($alias, '[]');
+                $function->addParamPhpDoc($alias, ArrayShapeGenerator::generate($configuration->getConfigTreeBuilder()->buildTree(), ArrayShapeGenerator::FORMAT_PHPDOC));
+                $function->addParamAttribute($alias, ArrayShapeGenerator::generate($configuration->getConfigTreeBuilder()->buildTree(), ArrayShapeGenerator::FORMAT_JETBRAINS_ATTRIBUTE));
```
```suggestion
                $tree = $configuration->getConfigTreeBuilder()->buildTree();
                $function->addParamPhpDoc($alias, ArrayShapeGenerator::generate($tree, ArrayShapeGenerator::FORMAT_PHPDOC));
                $function->addParamAttribute($alias, ArrayShapeGenerator::generate($tree, ArrayShapeGenerator::FORMAT_JETBRAINS_ATTRIBUTE));
```

dunglas commented on src/Symfony/Component/Config/Builder/FunctionBuilder.php:
```diff
@@ -0,0 +1,152 @@
+            }
+        }
+
+        return strtr('<?php
```
NOWDOC?

alexandre-daubois replied:
Updated 👍

nicolas-grekas commented:
To anyone willing to see improved support for this updated PHP config format, please upvote this phpstorm issue ;)
https://youtrack.jetbrains.com/issue/WI-79484/ArrayShape-and-PHPDoc-shapes-on-param-doesnt-hint-well

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,214 @@
+
+            if ($node->hasDefaultValue() && !\is_array($defaultValue = $node->getDefaultValue())) {
+                $hasContent = true;
+                $comment .= 'Default: '.json_encode($defaultValue).'. ';
```
let's add  JSON_UNESCAPED_SLASHES, JSON_UNESCAPED_UNICODE and JSON_PRESERVE_ZERO_FRACTION flags

and no need for the final dot

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,214 @@
+            $shape .= \sprintf("%s'%s' => ", str_repeat(' ', 4 * $nestingLevel), $child->getName());
+
+            if ($child instanceof PrototypedArrayNode) {
+                $shape .= '['.$handleNode($child->getPrototype(), $nestingLevel).']';
```
it's be nicer to put the phpdoc comment just after the opening bracket, eg
```diff
-                  'peer_fingerprint' => [
+                  'peer_fingerprint' => [ /* Associative array: hashing algorithm => hash(es). */
                        'sha1' => 'mixed',
                        'pin-sha256' => 'mixed',
                        'md5' => 'mixed',
                    ], /* Associative array: hashing algorithm => hash(es). */
                    'peer_fingerprint' => [
                        'sha1' => 'mixed',
                        'pin-sha256' => 'mixed',
                        'md5' => 'mixed',
-                   ], /* Associative array: hashing algorithm => hash(es). */
+                   ],
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/FunctionBuilder.php:
```diff
@@ -0,0 +1,154 @@
+
+REQUIREUSE
+/**
+ * This class is automatically generated to help in creating a config.
```
```suggestion
 * This function is automatically generated to help in creating a config.
```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/CacheWarmer/ConfigBuilderCacheWarmer.php:
```diff
@@ -68,19 +69,30 @@ public function warmUp(string $cacheDir, ?string $buildDir = null): array
             }
         }
 
+        if (class_exists(ConfigFunctionAwareBuilderGeneratorInterface::class) && $configurations) {
+            $generator->buildConfigFunction($configurations);
```
this might miss some tests... ;)
```suggestion
        if ($generator instanceof ConfigFunctionAwareBuilderGeneratorInterface && $configurations) {
            $generator->buildConfigFunction($configurations)();
```

alexandre-daubois replied:
Indeed, tests added 👍

nicolas-grekas reviewed (commented):
This is missing pre-filled shapes for "services", "imports" and "parameters" tree, which are not defined by a bundle, so don't have a configuration definition.

Also, can you investigate how an IDE like vscode could add support for auto-completing shapes? who's in charge of the most used vscode extensions so that we can ping them?

nicolas-grekas commented:
I'm also wondering if using named arguments for the 1st level provides the best DX: the IDE cannot know we want to type the name of an argument quickly enough I fear. Whereas if we start typing `['`, the IDE has a very precise context now.

alexandre-daubois commented:
I like the named argument way of doing because it is a bit more convenient to Cmd/Ctrl+Click on the named argument to have the full doc at a glance, whereas the Cmd/Ctrl+Click isn't available for array indexes. If there are a lot of bundles enabled, being able to go to the "definition" of the named argument could really help

alexandre-daubois commented:
> Also, can you investigate how an IDE like vscode could add support for auto-completing shapes?

It seems ArrayShapes won't be supported at all in VS Code : https://github.com/bmewburn/vscode-intelephense/issues/1476. Phpdoc seems to be supported (but partially) by what seems to be most used PHP extension. However, there's not much activity here neither: https://github.com/bmewburn/vscode-intelephense/issues/249 😕

alexandre-daubois commented:
> This is missing pre-filled shapes for "services", "imports" and "parameters" tree, which are not defined by a bundle, so don't have a configuration definition.

Do you mean we should have some first-level arguments calls `services`, `imports` and `parameters` in the `config()` function? That could work, the chance of colliding with an existing extension seems very small.

nicolas-grekas commented:
Yes that's what I mean. This cannot collide with an existing extension, those are reserved for DI config (if an extension tries that, their problem :) )

nicolas-grekas commented:
I'd suggest opening a new issue for https://github.com/bmewburn/vscode-intelephense/issues/249#issuecomment-1860083667, telling about nested shapes also.

stof commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,215 @@
+                return 'array<'.self::doGeneratePhpDoc($prototype, $nestingLevel).'>';
+            }
+
+            return 'array<'.self::handleNodeType($prototype).'>';
```
the type of keys should also be documented here, depending on whether the node is expecting a map or no (i.e. whether `getKeyAttribute()` returns a non-null value)

stof commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,215 @@
+            $arrayShape .= str_repeat(' ', $nestingLevel * 4).self::dumpNodeKey($child).': ';
+
+            if ($child instanceof PrototypedArrayNode) {
+                $arrayShape .= 'array<'.$handleNode($child->getPrototype(), $nestingLevel).'>';
```
same here about the key type

stof commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,215 @@
+
+        $arrayShape = "array{\n";
+
+        $handleNode = function (NodeInterface $node, int $nestingLevel): string {
```
is there any reason to make this a closure rather than using a private static method (which would then also be reused earlier

stof commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,215 @@
+        foreach ($children as $child) {
+            $arrayShape .= str_repeat(' ', $nestingLevel * 4).self::dumpNodeKey($child).': ';
+
+            if ($child instanceof PrototypedArrayNode) {
```
instead of this special case for `PrototypedArrayNode` everywhere, why not making `$handleNode` able to support it (my suggestion of using a private static method will simplify this a lot, as recursive closures are more painful than recursive methods due to needing to capture the closure by reference)

alexandre-daubois commented:
> I'd suggest opening a new issue for [bmewburn/vscode-intelephense#249 (comment)](https://github.com/bmewburn/vscode-intelephense/issues/249#issuecomment-1860083667), telling about nested shapes also.

I'm having a look at it. 

> This is missing pre-filled shapes for "services", "imports" and "parameters" tree, which are not defined by a bundle, so don't have a configuration definition.

That's now added 🙂

stof commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,215 @@
+        if ($node instanceof IntegerNode) {
+            $type = 'int<%s, %s>';
+        } elseif ($node instanceof FloatNode) {
+            $type = 'float<%s, %s>';
```
float ranges are not supported by Psalm or phpstan.

stof commented on src/Symfony/Component/Config/Tests/Builder/Fixtures/AddToList.config_function_output.php:
```diff
@@ -0,0 +1,102 @@
+ *     resource: string,
+ *     type?: string|int|float|bool,
+ *     ignoreErrors?: string|int|float|bool,
+ * }>|non-empty-list<string> $imports
```
This is the wrong type for imports. The implementation allows mixing strings and array shapes in the same array. You should have a union type between the array shape and `string` in the value type of that array instead.

stof commented on src/Symfony/Component/Config/Tests/Builder/Fixtures/AddToList.config_function_output.php:
```diff
@@ -0,0 +1,102 @@
+        ],
+    ])] array $add_to_list = [], 
+    // imports:
+    #[ArrayShape([[
```
this array shape is wrong. It defines a shape with a top-level key `0`, not a type where each value satisfy an array shape (PHPStorm does not have an attribute to define a type like `array<K, V>`, which is **not** an array shape.)

alexandre-daubois replied:
Yes indeed, I'm not sure how we should handle this case. Maybe not using the ArrayShape attribute at all in this case?

stof commented on src/Symfony/Component/Config/Tests/Builder/Fixtures/AddToList.config_function_output.php:
```diff
@@ -0,0 +1,102 @@
+        'translator' => [
+            'fallbacks' => ['string|int|float|bool'],
+            'sources' => ['string|int|float|bool'],
+            'books' => [ /* Deprecated: The child node "books" at path "add_to_list.translator.books" is deprecated. looks for translation in old fashion way */
```
does this actually work in PHPStorm or does it need to use a nested ArrayShape object or a phpdoc-based type  declaration. Their readme makes me thing that `ArrayShape` expects an array mapping keys to types, without any special handling for providing arrays there to represent array types without using a string representation of the type.

alexandre-daubois replied:
It works well like this (if I get your question right). Here an example of it with the `$framework` array of the config() function:

![image](https://github.com/user-attachments/assets/c2f6adb2-5ac1-4efb-bc79-8cf37c18c941)

Types are broken though, but keys autocompletion of nested shapes is working with this syntax.

stof reviewed (commented):
I suggest removing the generation of the Jetbrains attribute, as I'm 99% sure that they are broken (assuming that the attribute works a lot differently than what their readme suggests)

stof commented:
note that phpstorm is able to read array shapes defined in phpdoc (as long as keys don't start with a `@` because it does not properly handle quoted literals containing a `@`, for which you can [vote on the ticket reporting it](https://youtrack.jetbrains.com/issue/WI-70979/Parsing-of-PHPDoc-produces-syntax-error-if-phpdoc-contains-some-tag-in-single-quotes))

nicolas-grekas commented:
A nice use case for the phpstomr ArrayShape is being able to add comments to the type descriptions.
But I agree it'd be nicer to be able to put such comments in the phpdoc-types.
I remember we already talked about this but cannot find the issue on phpstan to track progress in this. Does anyone?

stof commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,215 @@
+
+    private static function dumpNodeKey(NodeInterface $node): string
+    {
+        return $node->getName().($node->isRequired() ? '' : '?');
```
This implementation is incomplete. Some keys might require writing them quoted, to avoid issues. There are 2 cases that require quoting:

- keys starting with a `@`, because it would be parsed as the start of a new phpdoc tag (while both phpstan and psalm support having a `@` inside a string literal
- probably strings that include a quote character as they are the delimiter for string literals, but I haven't tested this to confirm (at least, I'm sure it is needed if the quote is at the beginning as this would start a string literal)
- keys that contain ASCII upper case letters or are equal to the name of a builtin type (i.e. `['bool', 'null', 'int', 'float', 'double', 'scalar']) because of https://github.com/vimeo/psalm/issues/10008 (phpstan does not require quoting in that case, but it supports quoting the name without issues). See https://github.com/async-aws/aws/pull/1518 for the logic used in async-aws to cover that case.

alexandre-daubois replied:
Tried multiple things for the quotes in keys but Phpstan doesn't seem to support them at all. One example of this: https://phpstan.org/r/6b1bfbbe-038f-4cca-a4fa-fe46942598df

alexandre-daubois commented:
@nicolas-grekas yes, I opened https://github.com/phpstan/phpstan/issues/11895, which was related to https://github.com/phpstan/phpdoc-parser/issues/184.

alexandre-daubois commented:
@stof I addressed most comments:
- Escaping some special array-keys with quotes when necessary ;
- Use `$node->getKeyAttribute()` to determine if a hashmap is expected
- Fixed float unsupported intervals
- Fixed `$imports` array shape.

Indeed `ArrayShape` doesn't support well our prototyped arrays so I'm not sure what to do with it... Should we try to create a logic that guess if the given node supports completely ArrayShape attribute? Truly having all comments inline in the attribute to describe each node is a killer feature to me. It would be too bad to drop it.

nicolas-grekas commented:
> having all comments inline in the attribute to describe each node is a killer feature to me

I agree, and we need to involve everyone that can help on the topic, like we did on the phpstan repo. I'm for dropping ArrayShape attributes as a target and describe precisely what's missing for ideal phpdoc.

On my list:
- add support for inline comments to phpstan/doc-parser
- add support for inline comments to phpstorm
- same for psalm?
- add autocompletion support to intelephence (with support for inline comments)

I would be nice also to be able to describe an argument on its declaration point. Is something like this this supported by phpstan/psalm/etc? If not, it'd be nice, making things more local:

```php
function (
    /** @param array{...} description */ $theArg
) {...}
```

Let's get all parties involved to make this happen. This will not only benefit Symfony but all the php community.

alexandre-daubois commented:
To keep track of related issues, here are the new ones created:

- Support inline comments in PHPDoc in PHPStorm : https://youtrack.jetbrains.com/issue/WI-80360/Support-for-inline-comments-in-array-shapes
- Support inline comments in VSCode : https://github.com/bmewburn/vscode-intelephense/issues/3168
- Support shapes on calling in VSCode : https://github.com/bmewburn/vscode-intelephense/issues/3169

nicolas-grekas commented:
Here’s what I suggest next:

- Create a branch to back up your current work.
- Remove support for ArrayShape.
- Split the config function into separate functions for each extension within the Symfony\Config namespace—e.g., `framework($config)`, `import($config)`, etc.—all in a single file (config.php).
- Generate inline comments right away—this feature is for the future; by generating them now, we’ll encourage the rest to update
- Instead of @param, use `@var`, it might be better for readability: `function framework(/** @var array{...} $config */ array $config)`
- For `services()`/`import()`, there’s no need to redefine them using nodes. A template with manually written PHPDoc will be much easier to maintain.

alexandre-daubois commented:
PR updated with your suggestion @nicolas-grekas, you can find an example [in the fixtures](https://github.com/symfony/symfony/pull/58771/files#diff-210dee50f3127599c29e0b12fae68cff0e42bba9a16513421b8b2cb92cb6ad2b).

stof commented:
@nicolas-grekas I'm not sure `@var` works fine for function arguments (the only case where we use them right now is on promoted properties, where they work because `@var` works in properties)

stof commented:
> * For `services()`/`import()`, there’s no need to redefine them using nodes. A template with manually written PHPDoc will be much easier to maintain.

or even just a file directly in the package that gets required directly instead of copying a template to the file being generated (there would be nothing dynamic in the template here)

nicolas-grekas commented:
> or even just a file directly in the package that gets required directly

:100: the builder would just generate the appropriate require of the file located in the vendor dir

alexandre-daubois commented:
The `config.php` file now has the following requires and the files are added to the package resources:

```php
require_once dirname(__DIR__, 5).'/vendor/symfony/config/Resource/imports.php';
require_once dirname(__DIR__, 5).'/vendor/symfony/config/Resource/services.php';
require_once dirname(__DIR__, 5).'/vendor/symfony/config/Resource/parameters.php';
```

yceruto commented:
Nice! just wondering what is the difference between this feature and the fluent php config introduced here https://github.com/symfony/symfony/pull/40600?

nicolas-grekas commented:
In addition to what @alexandre-daubois wrote:

The fluent php config format did not make it as a replacement for the yaml one in felex recipes because it comes with unique challenges, namely patching / dumping an existing file is next to impossible.

With php arrays, it becomes as trivial as patching yaml.

That's THE reason that made me push to explore this alternative.

yceruto commented:
Thanks for the detailed explanation. I understand now that it's a better alternative than the fluent php config.

sorry, another question arose after reading your comments and PR description: will PhpStorm be able to autodetect the `$this` scope here, or do we need to do something special to enable autocomplete for `$this->`?

alexandre-daubois commented:
Good question, actually this is a mistake you see in the PR description example. Because of how PHP files are loaded by the DI, you have access to `$env` and `$path` variables in your PHP file, but `$this` should not be accessible.

alexandre-daubois commented:
Support for inline comments by PhpStan:
- https://github.com/phpstan/phpdoc-parser/pull/264

ondrejmirtes commented:
Yeah, it's coming, it's now at the top of my todolist :)

dunglas commented:
Also, it's very important easy to dump or encode a config coming from any store (Kubernetes config, YAML or JSON file...) into a PHP array, while it's almost impossible with the fluent interface.

This makes apps more "cloud native".

alexandre-daubois commented:
Great news, thank you @ondrejmirtes for being that responsive on this request!

fabpot commented on src/Symfony/Bundle/FrameworkBundle/CacheWarmer/ConfigBuilderCacheWarmer.php:
```diff
@@ -68,19 +69,31 @@ public function warmUp(string $cacheDir, ?string $buildDir = null): array
+                continue;
+            }
+
+            $alias = lcfirst(str_replace('_', '', ucwords($extension->getAlias(), '_')));
```
I think functions should use underscore names like in PHP, so `web_profiler()` and not `webProfiler()`.

alexandre-daubois replied:
I agree, this is how `WebProfiler` configuration function now looks like:

```php
function web_profiler(
    /** @var array{
     *     toolbar?: array{ // Profiler toolbar configuration
     *         enabled?: bool, // Default: false
     *         ajax_replace?: bool, // Replace toolbar on AJAX requests Default: false
     *     },
     *     intercept_redirects?: bool, // Default: false
     *     excluded_ajax_paths?: string|int|float|bool, // Default: "^/((index|app(_[\\w]+)?)\\.php/)?_wdt"
     * } $config */
    array $config = []): \Closure {
    return static function (\Symfony\Config\WebProfilerConfig $webProfilerConfig) use ($config) {
        $webProfilerConfig->configure($config);
    };
}
```

alexandre-daubois commented:
Comments in array shapes are live in the latest PhpStorm 2025.1 Beta:

![image](https://github.com/user-attachments/assets/6e2e91e8-2aed-4ae5-957c-42bec4cf9248)

nicolas-grekas commented:
I'm shaking again this config format to find the best syntax. I'm wondering about how one could define several arrays in the same file. We do so currently when some settings depends on the env - with `#[WhenEnv]`. Porting these to the new format is going to be not fun.

What about using a config object with as many methods as needed, and no return value?

```php
<?php

/** @var Symfony\Config\Config $config*/;

$config->framework([...]);

if ('dev' === $config->env) {
    $config->framework([...]);
}
```

If we want to keep splitting shapes in separate files (to ease scrolling the shape?) we might generate traits and use them in the generated Config class.

WDYT?

alexandre-daubois commented:
Suits me fine, I particularly like the trait part. I definitely think it would be _way_ easier to scroll through shapes if we have `FrameworkConfigTrait`, `WebProfilerConfigTrait`, etc.

> What about using a config object with as many methods as needed, and no return value

I agree, just a suggestion on the last bit. What about actually returning `$this` to chain this way?

```php
$config
    ->framework([...])
    ->webProfiler([...])
    ->security([...]);
```

nicolas-grekas commented:
JetBrains / phpStorm did their part, thanks to them!

> [#[ArrayShape] and PHPDoc shapes on param doesn't hint well : WI-79484](https://youtrack.jetbrains.com/issue/WI-79484/ArrayShape-and-PHPDoc-shapes-on-param-doesnt-hint-well)

> [Allow comments in array shape : WI-73436](https://youtrack.jetbrains.com/issue/WI-73436/Allow-comments-in-array-shape)

> What about actually returning $this to chain this way?

Why not indeed.

Up to resume work on this topic?

alexandre-daubois commented:
It's a bit of a rush at the moment, but as soon as I have a bit of time to get on with the job, this PR is my priority. 🙂

alexandre-daubois commented:
I just pushed the revamped version. PR description updated with the new syntax. After playing with, navigating through the shapes is much simpler now that traits are generated.

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,158 @@
+
+    private static function generateInlinePhpDocForNode(BaseNode $node): string
+    {
+        $hasContent = false;
```
needless var: $comment should start empty and be prefixed only at the end, if not empty anymore
also, CR/LF in $comment should be replaced by a space so that we're sure it remains a oneliner

alexandre-daubois replied:
It outputs something like this:

```
     *         strict_requirements?: string|int|float|bool, // set to true to throw an exception when a parameter does not match the requirements set to false to disable exceptions when a parameter does not match the requirements (and return null instead) set to null to disable parameter checks against requirements 'true' is the preferred configuration in development mode, while 'false' or 'null' might be preferred in production Default: true
```

Maybe we should replace `/\s/` with `; ` instead?

alexandre-daubois replied:
It currently outputs something like this for `framework`:

```
     *         strict_requirements?: string|int|float|bool, // set to true to throw an exception when a parameter does not match the requirements set to false to disable exceptions when a parameter does not match the requirements (and return null instead) set to null to disable parameter checks against requirements 'true' is the preferred configuration in development mode, while 'false' or 'null' might be preferred in production Default: true
```

Maybe we could instead replace `\r` and `\n` with `; `?

alexandre-daubois replied:
Actually maybe not, it seems it's a particular case. Other places end the info with a `.`, so maybe we should adjust the info of this particular node.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/PhpFileLoader.php:
```diff
@@ -57,16 +60,38 @@ public function load(mixed $resource, ?string $type = null): mixed
+
         // the closure forbids access to the private scope in the included file
-        $load = \Closure::bind(function ($path, $env) use ($container, $loader, $resource, $type) {
+        $config = class_exists(Config::class) ? new Config() : null;
```
Why does the generated Config class have constructor arguments when those are never used?
Also, it'd be nice to add public readonly vars to the $config object to get access to the other tools one could use, especially `$config->env` (possibly $config->path, container, loader, resource, type, although I'm not sure what they could be used for there - for thoughts.)

alexandre-daubois replied:
Updated, it now looks like this:

```php
<?php

namespace Symfony\Config;

use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
use Symfony\Config\FrameworkConfig;
use Symfony\Config\SecurityConfig;
use Symfony\Config\TwigConfig;
use Symfony\Config\WebProfilerConfig;
use Symfony\Config\MonologConfig;
use Symfony\Config\DebugConfig;

/**
 * This class is automatically generated to help in creating a config.
 */
class Config 
{
    use \Symfony\Component\DependencyInjection\Resource\ImportsTrait;
    use \Symfony\Component\DependencyInjection\Resource\ParametersTrait;
    use \Symfony\Component\DependencyInjection\Resource\ServicesTrait;
    use \Symfony\Config\FrameworkTrait;
    use \Symfony\Config\SecurityTrait;
    use \Symfony\Config\TwigTrait;
    use \Symfony\Config\WebProfilerTrait;
    use \Symfony\Config\MonologTrait;
    use \Symfony\Config\DebugTrait;
    private $builders;
    private $frameworkConfig;
    private $securityConfig;
    private $twigConfig;
    private $webProfilerConfig;
    private $monologConfig;
    private $debugConfig;
    public function __construct(public readonly ?string $env)
    {
        $this->frameworkConfig = new FrameworkConfig();
        $this->securityConfig = new SecurityConfig();
        $this->twigConfig = new TwigConfig();
        $this->webProfilerConfig = new WebProfilerConfig();
        $this->monologConfig = new MonologConfig();
        $this->debugConfig = new DebugConfig();
        $this->builders = [$this->frameworkConfig, $this->securityConfig, $this->twigConfig, $this->webProfilerConfig, $this->monologConfig, $this->debugConfig];
    }
    public function getBuilders(): array
    {
        return $this->builders;
    }

}
```

`$env` is injected. It's always possible to add other ones if needed, `$env` being the most important to me as well

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Resource/ParametersTrait.php:
```diff
@@ -0,0 +1,32 @@
+    private array $parameters = [];
+
+    /**
+     * @var array<string, mixed> $parameters
```
isn't mixed too broad?

alexandre-daubois replied:
Narrowed the type a bit to `array<string, array|bool|string|int|float|\UnitEnum|null>`

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Resource/ServicesTrait.php:
```diff
@@ -0,0 +1,35 @@
+     * @var array<array{
+     *     id: string,
+     *     class?: string|int|float|bool,
+     * }>|array<string, class-string|null> $services
```
This is minimalistic. In reality, we could be way more descriptive here, see what YamlFileLoader describes for services. Maybe you preferred postponing this to a follow up?

alexandre-daubois replied:
I started implementing what's left (lazy, abstract, args, synthetic, etc.) but there's a bit more work than I expected. I propose to validate the PR as-is, then comeback to this bit when the rest is OK?

alexandre-daubois committed:
Fulfil services

alexandre-daubois commented:
I completed the services array shape in f1857324c1a509254ae040516dbb9b3346f41cdf.

nicolas-grekas commented:
I feel like we might have too much indirection with the current approach.
I'm also not sure the required `@var` will provide the best DX.
Another approach that might overcome those issues: have a syntax like this:

```php

namespace Symfony\Config;

return new FrameworkConfig([...]);
```

IIUC, the constructor is already accepting the arrays we want, so we should just have to remove the current `configure()` method and move the shape on the constructor, isn't it?

And in order to be able to configure many trees at once, we could also support returning an iterable:

```php

namespace Symfony\Config;

return [
new FrameworkConfig([...]),
new ServicesConfig([...]),
];
```

WDYT? Wouldn't this make things a bit simpler?

alexandre-daubois commented:
What do you mean by "too much indirection"?

> IIUC, the constructor is already accepting the arrays we want, so we should just have to remove the current configure() method and move the shape on the constructor, isn't it?

Makes sense indeed, let me have a look at this

> Wouldn't this make things a bit simpler?

I find the current way already really nice. However, I don't have a strong opinion on this and the syntax you propose is also fine to me. Maybe it will finish to convince me once I understand the "indirection" part 🙂

nicolas-grekas commented:
Indirection from the Config class to the traits to the configure() methods.
And repetition of the shape on the traits and on configure() methods also.

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,153 @@
+            $node instanceof StringNode => 'string',
+            $node instanceof NumericNode => self::handleNumericNode($node),
+            $node instanceof EnumNode => $node->getPermissibleValues('|'),
+            $node instanceof ScalarNode => 'string|int|float|bool',
```
```suggestion
            $node instanceof ScalarNode => 'string|int|float|bool|null',
```

alexandre-daubois commented:
I think that the task is now complete. Many PRs were merged lately about this topic, so this one isn't necessary anymore. The feature will be shipped in 7.4. Many thanks to everyone involved! 🚀
